Utils/Substitution.hs

```diff
@@ -347,11 +349,29 @@ instance FreshenLclBndrs (FcTerm a) where
               <*> freshenLclBndrs (substVar x (FcTmVar y) tm1)
               <*> freshenLclBndrs (substVar x (FcTmVar y) tm2)
 
-  freshenLclBndrs (FcTmCase tm cs) = FcTmCase <$> freshenLclBndrs tm <*> mapM freshenLclBndrs cs
+  freshenLclBndrs (FcTmCase tm cs)   = FcTmCase <$> freshenLclBndrs tm <*> mapM freshenLclBndrs cs
+  freshenLclBndrs (FcTmCaseNs tm cs) = FcTmCaseNs <$> freshenLclBndrs tm <*> mapM freshenLclBndrs cs
 
 -- | Freshen the (type + term) binders of a System F case alternative
 instance FreshenLclBndrs (FcAlt a) where
-  freshenLclBndrs (FcAlt (FcConPat dc xs) tm) = do
-    ys  <- mapM (\_ -> freshFcTmVar) xs
-    tm' <- freshenLclBndrs $ foldl (\t (x,y) -> substVar x (FcTmVar y) t) tm (zipExact xs ys)
-    return (FcAlt (FcConPat dc ys) tm')
+  freshenLclBndrs (FcAlt p tm) = do
+    (p', subst) <- freshenPatLclBndrs p
+    tm' <- freshenLclBndrs (subst tm)
+    return (FcAlt p' tm')
+
+
+freshenPatLclBndrs :: MonadUnique m => FcPat a -> m (FcPat a, FcTerm a -> FcTerm a)
+freshenPatLclBndrs (FcConPat dc xs) = do
+  ys  <- mapM (\_ -> freshFcTmVar) xs
+  return (FcConPat dc ys, foldr (.) id (map (\(x, y) -> substVar x (FcTmVar y)) (zipExact xs ys)))
```

Comment:
Best if you split up this big line. Something like this should do it:
```
freshenPatLclBndrs (FcConPat dc xs) = do
  ys  <- mapM (\_ -> freshFcTmVar) xs
  let subst = foldr (\(x,y) s -> substVar x (FcTmVar y) . s) id (zipExact xs ys)
  return (FcConPat dc ys, subst)
```

Reply:
Done 😄